Add support for codecov.io

leandro-lucarella-sociomantic opened the pull request:
We should move the [modified](https://github.com/sociomantic-tsunami/ocean/blob/v2.x.x/ci/codecov.sh) [codecov-bash](https://github.com/codecov/codecov-bash) script here until the [upstream PR](https://github.com/codecov/codecov-bash/pull/72) gets merged. We should also move the [common reports and config gathering and running the script in a docker container](https://github.com/sociomantic-tsunami/ocean/blob/fecdcefaa53277ff2e565958da29267f7520fc1b/ci/travis-after-success.sh#L4-L16) here too.

leandro-lucarella-sociomantic commented:
Updated script due to problems with not having a git repo: https://github.com/sociomantic-tsunami/ocean/pull/134

leandro-lucarella-sociomantic commented:
Codecov issue was merged. FYI @mihails-strasuns-sociomantic

leandro-lucarella-sociomantic committed:
dlang: Add command to send reports to codecov

Fixes #1.

leandro-lucarella-sociomantic commented:
Code attached.

leandro-lucarella-sociomantic commented:
Test PR in ocean: https://github.com/sociomantic-tsunami/ocean/pull/231

leandro-lucarella-sociomantic committed:
dlang: Pass a few extra Makd env variables to make

These variables are very widely used, so it makes sense to be passed by
default. This helps to mitigate #25 but at some point we need a more
reliable solution.

nemanja-boric-sociomantic commented on README.md:
```diff
@@ -347,3 +347,25 @@ The `beaver dlang make` command also uses some utilities that could come handy
+
+Only `TRAVIS*`, `CODECOV_*` and some dlang-related (`DIST DMD DC F V` etc.)
+environment variables will be, if you want to pass any other environment
+variables use `BEAVER_DOCKER_VARS` as usual.
```
I get what this sentence is saying, but after reading it for the second time.


> Only `TRAVIS*`, `CODECOV_*`, and some dlang-related (`DIST DMD DC F V` etc.) environment variables will be passed to a running environment. If you want to pass any other environment variable use `BEAVER_DOCKER_VARS` as usual.

leandro-lucarella-sociomantic replied:
This was definitely a half-edit that got truncated :D

nemanja-boric-sociomantic commented on README.md:
```diff
@@ -347,3 +347,25 @@ The `beaver dlang make` command also uses some utilities that could come handy
+environment variables will be, if you want to pass any other environment
+variables use `BEAVER_DOCKER_VARS` as usual.
+
+The following environment variables are reported to codecov (via the `-e`
```
> The following environment variables are automatically (via the `-e` option) reported to code cov: `DIST`, `DMD`, `DC`, `F`.

leandro-lucarella-sociomantic replied:
Mmm, I don't like it this way, take a look at my new take and let me know what you think.

Burgos replied:
Yeah, I think it's fine.

mihails-strasuns-sociomantic commented on bin/dlang/codecov:
```diff
@@ -0,0 +1,37 @@
+codecov_vars="$(printenv -0 | sed -zn 's/^\(CODECOV_[^=]\+\)=.*$/\1\n/p' |
+        tr -d '\0')"
+# Export D related environment variables
+dlang_vars="DIST DMD DC DVER D2_ONLY F V"
```
Shouldn't be needed at all, codecov script only works with generated .lst file and git history

leandro-lucarella-sociomantic replied:
It's metadata, and I think is nice to know how the run was built to come up with that coverage. You can see it here for example (codecov sends `TRAVIS_OS` by default): https://codecov.io/gh/sociomantic-tsunami/ocean/commit/73ee571debc6e399e7acc465fc6dd28f9afe2cd7/build

We could also use labels for some things, like `F` and `DIST` could be good candidates, maybe even `DC`. (I don't know how they look, so I don't know how useful they would be visually). I might go for a try...

mihails-strasuns-sociomantic replied:
OK, if you intend to submit it as part of metadata some way, it makes sense to me.

leandro-lucarella-sociomantic replied:
Yup, it's done, have a look at ocean's PR :)

mihails-strasuns-sociomantic commented on README.md:
````diff
@@ -347,3 +347,25 @@ The `beaver dlang make` command also uses some utilities that could come handy
+then use something like this in the `.travis.yml` file:
+
+```yml
+after_success: beaver dlang codecov [OPTIONS]
````
If I understand correctly, the script auto-detects languages so it should be just plain `beaver codecov`.

leandro-lucarella-sociomantic replied:
It has many option, like sending env vars, sending labels, enabling or disabling features, I can already see people wanting to use those.

mihails-strasuns-sociomantic replied:
?

All I was saying that it is semantically `beaver codecov [OPTIONS]`, not `beaver dlang codecov [OPTIONS]` because it will work with any supported code cov language.

leandro-lucarella-sociomantic replied:
But this script only works with D projects, it assuming a lot of stuff that work only with them (like that report files have the pattern `.*.lst`, and it passes the `F` and other Makd env vars).

We need something specific for D for now, so I prefer to limit to it to simplify things and get the most out of this setup.

Burgos commented on README.md:
```diff
@@ -347,3 +347,26 @@ The `beaver dlang make` command also uses some utilities that could come handy
+you want to pass any other environment variables use `BEAVER_DOCKER_VARS` as
+usual.
+
+The following environment variables are reportedautomatically to codecov (via
```
Missing space.

leandro-lucarella-sociomantic replied:
Fixed

leandro-lucarella-sociomantic committed:
dlang make: Upload codecov reports automatically

This way we can assign appropriate labels (flags) to the reports being
uploaded, which improves visualization.

leandro-lucarella-sociomantic commented:
I think this is ready for a final review + merge. Seems to be working nicely with Ocean.

mihails-strasuns-sociomantic commented on bin/dlang/make:
```diff
@@ -7,28 +7,42 @@
+    rm -f .*.lst
+}
+test_cov unittest
+test_cov integrationtest
```
Until we have a way to output coverage reports from different integration tests to different folders and submit them all, I would recommend disabling this part completely. Right now you will get misleading flakey reports from whatever integration tests has run last.

leandro-lucarella-sociomantic replied:
OK, good point.

leandro-lucarella-sociomantic replied:
https://github.com/sociomantic-tsunami/makd/issues/107

leandro-lucarella-sociomantic replied:
Updated.

leandro-lucarella-sociomantic commented:
Ready for a final review

mihails-strasuns-sociomantic commented:
Red

leandro-lucarella-sociomantic commented:
:heart: tests

leandro-lucarella-sociomantic commented:
(I forgot to call `integrationtests` without sending coverage reports after I commented out the helper function to send the reports when making `integrationtest`)

leandro-lucarella-sociomantic commented:
:crossed_fingers:

leandro-lucarella-sociomantic commented:
:green_heart:

leandro-lucarella-sociomantic commented:
ping @mihails-strasuns-sociomantic
